src/aws-cpp-sdk-core/include/aws/core/utils/stream/AwsChunkedStream.h

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+#include <aws/core/utils/HashingUtils.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/crypto/Hash.h>
+
+namespace Aws {
+namespace Utils {
+namespace Stream {
+
+static const size_t AWS_DATA_BUFFER_SIZE = 65536;
+
+template <size_t DataBufferSize = AWS_DATA_BUFFER_SIZE>
+class AwsChunkedStream {
+ public:
+  AwsChunkedStream(Http::HttpRequest *request, const std::shared_ptr<Aws::IOStream> &stream)
+      : m_chunkingStream{Aws::MakeShared<StringStream>("AwsChunkedStream")}, m_request(request), m_stream(stream) {}
+
+  size_t BufferedRead(char *dst, size_t amountToRead) {
+    // the chunk has ended and cannot be read from
+    if (m_chunkEnd) {
+      return 0;
+    }
+
+    // If we've read all of the underlying stream write the checksum trailing header
+    // the set that the chunked stream is over.
+    if (m_stream->eof() && (m_chunkingStream->eof() || m_chunkingStream->peek() == EOF)) {
+      Aws::StringStream chunkedTrailer;
+      chunkedTrailer << "0\r\n";
+      if (m_request->GetRequestHash().second != nullptr) {
+        chunkedTrailer << "x-amz-checksum-" << m_request->GetRequestHash().first << ":"
+                       << HashingUtils::Base64Encode(m_request->GetRequestHash().second->GetHash().GetResult()) << "\r\n";
+      }
+      chunkedTrailer << "\r\n";
+      auto trailerSize = chunkedTrailer.str().size();
+      memcpy(dst, chunkedTrailer.str().c_str(), trailerSize);
+      m_chunkEnd = true;
+      return trailerSize;
+    }
+
+    // Try to read in a 64K chunk, if we cant we know the stream is over
+    size_t bytesRead = 0;
+    while (m_stream->good() && bytesRead < DataBufferSize) {
+      m_stream->read(&m_data[bytesRead], DataBufferSize - bytesRead);
+      bytesRead += m_stream->gcount();
+    }
+
+    // update the trailing checksum to be sent only if we read data and buffered.
+    if (bytesRead > 0 && m_request->GetRequestHash().second != nullptr) {
+      m_request->GetRequestHash().second->Update(reinterpret_cast<unsigned char *>(m_data.data()), bytesRead);
+    }
+
+    // Buffer chunked encoding from data if there was data read to the buffer, otherwise leave it alone/
+    if (bytesRead > 0 && m_chunkingStream != nullptr) {
+      *m_chunkingStream << Aws::Utils::StringUtils::ToHexString(bytesRead) << "\r\n";
+      std::copy(m_data.begin(), m_data.begin() + bytesRead, std::ostream_iterator<char>(*m_chunkingStream));
+      *m_chunkingStream << "\r\n";
+      auto curr = m_chunkingStream->tellg();
+      const auto rn = m_chunkingStream->rdbuf();
+      AWS_UNREFERENCED_PARAM(rn);
```

Comment:
why do we need rn?

Reply:
ooooooooo, that was a debugging variable, oooooooops